nixos/acme: Fixes for account creation and remove tmpfiles usage

m1cr0man opened the pull request:
###### Motivation for this change

Closes #106565
Possibly closes some other ACME tickets related to JWS token errors too. A new systemd target named `acme-account-$hash.target` has been added, which is depended on (requiredBy + after) by all but 1 cert service associated with the same account. The one remainder is depended on (requires + before) to do the account creation.
Note that I changed the hash data for the account directory which will trigger new account creation on all systems. Given that rate limit issues should now be resolved, this will simply test that I did a good enough job of it ;)

Closes #106603
Umask is changed to 0023 to accommodate for lighttpd

Closes #103121
Remove systemd-tmpfiles dependency follows from [this discussion](https://github.com/NixOS/nixpkgs/pull/97702#issuecomment-733963047) and ~may solve~ has been reported to solve #103121. I was tempted to totally remove the use of tmpfiles but I couldn't find a good home for the webroot directory creation. Ideally, it should already be created by the user but since most users don't set a webroot at all and expect it to "just work" through httpd/nginx I left it in.

Docs update are primarily to cover follow-up in #81634 and fix the `systemctl clean` command.

In the test suite, I added a wait for "network-online.target" on the pebble server due to a one time test failure I had because the host wasn't reachable. I hope this resolves it, but it's hard to test.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

m1cr0man committed:
nixos/acme: prevent mass account creation

Closes #106565
When generating multiple certificates which all
share the same server + email, lego will attempt
to create an account multiple times. By adding an
account creation target certificates which share
an account will wait for one service (chosen at
config build time) to complete first.

m1cr0man committed:
nixos/acme: reduce dependency on tmpfiles

systemd-tmpfiles is no longer required for
most of the critical paths in the module. The
only one that remains is the webroot
acme-challenge directory since there's no
other good place for this to live and forcing
users to do the right thing alone will only
create more issues.

m1cr0man committed:
nixos/acme: change service umask to 0023

Closes #106603
Some webservers (lighttpd) require that the
files they are serving are world readable. We
do our own chmods in the scripts anyway, and
lego has sensible permissions on its output
files, so this change is safe enough.

m1cr0man committed:
nixos/acme: update documentation and release notes

The instructions on recreating the cert were missing --what=state.
Also added a note on ensuring the group of manual certs is correct.

m1cr0man commented:
Heads up @NixOS/acme

m1cr0man committed:
nixos/acme: fix test config

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/lets-encrypt-on-20-09/9950/13

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/migrating-letsencrypt-acme-certificates-from-another-server/9925/2

symphorien commented:
I didn't read this in detail, but I wonder if using a systemd target is the right thing. Starting a target twice will not start the dependent services twice if I understand correctly.

m1cr0man commented:
Rebased

flokli commented:
This is now missing all the cherry-picked-by lines (did it miss them before?).

Backport PRs should always be cherry-picked with the `-x` parameter.

m1cr0man commented:
Uh, this isn't a backport? These PRs are fresh. I accidently cherry-picked and pushed your commit for #107769 in here, then reverted it.

symphorien commented on nixos/modules/security/acme.nix:
```diff
@@ -222,10 +231,14 @@ let
+        # Keep in mind that these directories will be deleted if the user runs
+        # systemctl clean --what=state
+        # acme/.lego/${cert} is listed for this reason.
+        StateDirectory =
```
(nitpick) I think a list of string would have the same effect.

m1cr0man replied:
This has the effect of changing the generated entries in the unit config from a single line of space separated values to multiple lines for each value. The resulting behaviour is the same, so I changed it anyway in favour of the Nix syntax improvement.

m1cr0man commented:
> I didn't read this in detail, but I wonder if using a systemd target is the right thing. Starting a target twice will not start the dependent services twice if I understand correctly.

This is correct, however it is not an issue in this case. The target is only required to solve issues on that first start, where the account directory doesn't exist. Subsequent renewals won't have any issues so long as this target has been reached at least once.

symphorien commented:
Right, I understand.

arianvp commented:
> I didn't read this in detail, but I wonder if using a systemd target is the right thing. Starting a target twice will not start the dependent services twice if I understand correctly.

It will  if the service is down. Systemd will "approach" the target

Try stopping for example `nscd.service` and then do `systemctl start multi-user.target`. It will start the service again

m1cr0man committed:
nixos/acme: Remove all systemd-tmpfiles usage

- Added an ExecPostStart to acme-$cert.service when webroot is defined to create the acme-challenge
directory and fix required permissions. Lego always tries to create .well-known and acme-challenge,
thus if any permissions in that tree are wrong it will crash and break cert renewal.
- acme-fixperms now configured with acme User and Group, however the script still runs as root. This
ensures the StateDirectories are owned by the acme user.
- Switched to list syntax for systemd options where multiple values are specified.

m1cr0man commented:
I applied this config to my own system and encountered some new permissions issues with the first renewal, which I've now fixed:

- The accounts directory (and subdirectories, oddly) was being set to root:root ownership because acme-fixperms.service was running as root, thus the StateDirectories were being set to root ownership. I changed the service to use the acme user and group, and kept the script running as root to fix erroneous permissions.
- I discovered that if the webroot is not owned by the user/group configured for cert renewal, Lego will crash due to being unable to mkdir .well-known and/or .well-known/acme-challenge. I added an ExecPreStart to the acme-$cert.service to fix this on launch. In the process I removed all usages of systemd-tmpfiles which is nice.

The commit also contains some other syntax fixes.

hlolli commented:
I had very annoying service restart problem with nginx using acme. disabling security/acme.nix module and import these changes fixes it for me.

symphorien commented on nixos/modules/security/acme.nix:
```diff
@@ -54,23 +59,34 @@ let
+  userMigrationService = let
     script = with builtins; concatStringsSep "\n" (mapAttrsToList (cert: data: ''
-      for fixpath in /var/lib/acme/${escapeShellArg cert} /var/lib/acme/.lego/${escapeShellArg cert}; do
+      chown -R acme .lego/accounts
```
this can go outside mapAttrsToList

symphorien commented on nixos/modules/security/acme.nix:
```diff
@@ -664,21 +695,25 @@ in {
+          leader = "acme-${(builtins.head confs).cert}.service";
+          dependantServices = map (conf: "acme-${conf.cert}.service") (builtins.tail confs);
+        in nameValuePair "acme-account-${hash}" {
+          requiredBy = dependantServices;
+          before = dependantServices;
+          requires = [ leader ];
+          after = [ leader ];
```
The same code but with comments because it took me a loooong time to understand

```suggestion
          # first, only run the service of one single certificate with this account
          # it will create the account
          requires = [ leader ];
          after = [ leader ];
          # The other services of this account can only run afterwards.
          requiredBy = dependantServices;
          before = dependantServices;
```

Maybe requiredBy is too strong: if the leader fails for an unrelated reason, won't it prevent renewal of other certs ? Unless targets are once again special ?

m1cr0man replied:
Yeah, they will fail, and I think that's fine. It could fail for reasons relating to creating an account, and you wouldn't want all your other cert services to then hammer LetsEncrypt and get the same errors (and subsequently get rate limited). I can understand a situation where the leader fails for non-account related reasons (domain ownership, web server config, etc) but again it should be intervened by the user at that point.

m1cr0man commented on nixos/modules/security/acme.nix:
```diff
@@ -54,23 +59,34 @@ let
     script = with builtins; concatStringsSep "\n" (mapAttrsToList (cert: data: ''
-      for fixpath in /var/lib/acme/${escapeShellArg cert} /var/lib/acme/.lego/${escapeShellArg cert}; do
+      chown -R acme .lego/accounts
+      for fixpath in ${escapeShellArg cert} .lego/${escapeShellArg cert}; do
```
No, and the next line checks that. The purpose of this service is to fix the permissions of existing data if/when configuration is changed rather than make it right on creation (which systemd + StateDirectory does just fine)

symphorien replied:
Ah right, I had not seen it :°

m1cr0man commented on nixos/modules/security/acme.nix:
```diff
@@ -244,7 +267,16 @@ let
+        ExecStartPre = "+" + pkgs.writeShellScript "acme-${cert}-make-webroot" ''
+          mkdir -p '${data.webroot}/.well-known/acme-challenge'
+          cd '${data.webroot}'
+          chown 'acme:${data.group}' . .well-known .well-known/acme-challenge
```
Yeah, that's a good idea. I'll give it a shot and see what happens

m1cr0man replied:
Ok I ran into an issue implementing this. In the acme test suite, the configured webroot is `/var/lib/acme/.acme-challenge`, which is [the default](https://github.com/NixOS/nixpkgs/blob/9622a8981bc71964b6364bbf3115c65b2352875c/nixos/modules/services/web-servers/nginx/vhost-options.nix#L72). With the removal of tmpfiles, neither nginx nor the acme module create this folder if it does not exist. This is a problem because systemd will not create source folders for bind paths if they do not exist, generating the following error (which is misleading... it _is_ a source directory problem):
```
webserver # [   12.092725] systemd[872]: acme-a.example.test.service: Failed to set up mount namespacing: /run/systemd/unit-root/tmp/webroot: No such file or directory
```
Note that the preStart script I wrote uses `mkdir -p` which circumvents this issue. I also tried an inverse symlink, where I created a pseudo-webroot inside `/var/lib/acme/.lego/${cert}/webroot`, and then linked it to the service's `/tmp` folder, and then linked the acme-challenge folder from the pseudo-webroot into the real webroot. I encountered some 404's 
during renewal which I believe were due to permission errors, but I didn't investigate further as I realised this would not work if more than one cert shared the same webroot.

In terms of chowning webroot, I gave this a bit of thinking and came up with a better solution. If the webroot should be owned by acme, then it should have permission to create it too. This is true if it resides in /var/lib/acme, as it does by default. I moved this script into the main start script and removed the chown of webroot. I kept chown for the other 2 folders in case the configured group changes (e.g. switching web servers).

m1cr0man committed:
nixos/acme: Incorporate review suggestions

m1cr0man committed:
nixos/acme: Set up webroot as non-root user

m1cr0man committed:
nixos/acme: Fix bash issue, enable debug

I found a logical error in the bash script, but during
debugging I enabled command echoing and realised it
would be a good idea to have it enabled all the time for
ease of bug reporting.

ryantm commented:
This worked for me.

Gaelan commented:
Anything that needs to be done before this is merged?

m1cr0man commented:
Nope, I don't have any more changes to make to this.

ryantm commented:
Can we get some more approvals from the ACME team?

@aanderse @andrew-d @arianvp @emilazy @flokli

m1cr0man commented:
I would also like to highlight the other smaller acme PRs I've had open for months. They too need to be merged for some essential fixes. At this point, I would prefer to see this PR merged first then I can work on rebasing and adjusting the other PRs as necessary.

aanderse commented:
> Can we get some more approvals from the ACME team?
> 
> @aanderse @andrew-d @arianvp @emilazy @flokli

A number of people have reported that this PR fixes their issues. The proof is in the pudding, so I recommend merge :+1:

flokli commented:
Let's get this in.

rubenmoor commented:
I still get the 403, using nixops on an AWS EC2 instance.

    nix-shell> nixops deploy -d staging
    
    main............> Jan 28 22:11:30 serendipity acme-staging.rubenmoor.net-start[17863]: 2021/01/28 22:11:30 [INFO] [staging.rubenmoor.net] acme: Trying to solve HTTP-01
    main............> Jan 28 22:11:36 serendipity acme-staging.rubenmoor.net-start[17863]: 2021/01/28 22:11:36 [INFO] Deactivating auth: https://acme-v02.api.letsencrypt.org/acme/authz-v3/10432246082
    main............> Jan 28 22:11:36 serendipity acme-staging.rubenmoor.net-start[17863]: 2021/01/28 22:11:36 [INFO] Unable to deactivate the authorization: https://acme-v02.api.letsencrypt.org/acme/authz-v3/10432246082
    main............> Jan 28 22:11:36 serendipity acme-staging.rubenmoor.net-start[17863]: 2021/01/28 22:11:36 Could not obtain certificates:
    main............> Jan 28 22:11:36 serendipity acme-staging.rubenmoor.net-start[17863]:         error: one or more domains had a problem:
    main............> Jan 28 22:11:36 serendipity acme-staging.rubenmoor.net-start[17863]: [staging.rubenmoor.net] acme: error: 403 :: urn:ietf:params:acme:error:unauthorized :: Invalid response from http://staging.rubenmoor.net/.well-known/acme-challenge/gLk9Zwwlz1RTXqPBWyOC2i0xuQTDaLtd7nUk89Fl6WA [54.93.104.227]: "<html>\r\n<head><title>403 Forbidden</title></head>\r\n<body>\r\n<center><h1>403 Forbidden</h1></center>\r\n<hr><center>nginx</center>\r\n", url:
    main............> Jan 28 22:11:36 serendipity systemd[1]: acme-staging.rubenmoor.net.service: Main process exited, code=exited, status=1/FAILURE
    main............> Jan 28 22:11:36 serendipity systemd[1]: acme-staging.rubenmoor.net.service: Failed with result 'exit-code'.
    main............> Jan 28 22:11:36 serendipity systemd[1]: Failed to start Renew ACME certificate for staging.rubenmoor.net.
    main............> Jan 28 22:11:36 serendipity systemd[1]: acme-staging.rubenmoor.net.service: Consumed 97ms CPU time, received 14.5K IP traffic, sent 7.3K IP traffic.
    
This is my `shell.nix` to point to a local clone of `nixpkgs`: https://github.com/m1cr0man/nixpkgs.git:master

```
let
  # acme bugfix on 20.09
  pkgsSrc = ~/nixpkgs;

  pkgs = (import pkgsSrc {});
in
pkgs.stdenv.mkDerivation rec {
  name = "nixops-env";
  buildInputs = with pkgs; [ awscli pkgs.nixops ];
  shellHook = ''
    [ -f ~/.aws/vars ] && . ~/.aws/vars
    export NIX_PATH=${pkgs.path}:nixpkgs=${pkgs.path}:.
  '';
}
```

And I can confirm that, upon visiting http://staging.rubenmoor.net/.well-known/acme-challenge/gLk9Zwwlz1RTXqPBWyOC2i0xuQTDaLtd7nUk89Fl6WA in a browser, that there is 403.

m1cr0man commented:
Assuming you're using the default webroot for certs, can you check the permissions and ownership of `/var/lib/acme/acme-challenges/.well-known/acme-challenge` from acme-challenges up to the end? I suspect, due to this https://github.com/NixOS/nixpkgs/pull/106857#discussion_r551013681, that your webroot permissions may be wrong. Please collect that and report back, we can work on a fix from there.

rubenmoor commented:
> Assuming you're using the default webroot for certs, can you check the permissions and ownership of `/var/lib/acme/acme-challenges/.well-known/acme-challenge` from acme-challenges up to the end? I suspect, due to this [#106857 (comment)](https://github.com/NixOS/nixpkgs/pull/106857#discussion_r551013681), that your webroot permissions may be wrong. Please collect that and report back, we can work on a fix from there.

Owner and group are `acme` for all of these folders 

```
/var/lib/acme/.challenges/.well-known/acme-challenge
/var/lib/acme/.challenges/.well-known/
/var/lib/acme/.challenges/
/var/lib/acme/
```

m1cr0man commented:
Ah of course, I'm just after reading your config thoroughly. Because you've manually configured your certs through nginx instead of using the builtins (like [services.nginx.virtualHosts.<name>.enableACME](https://search.nixos.org/options?channel=unstable&show=services.nginx.virtualHosts.%3Cname%3E.enableACME&from=0&size=50&sort=relevance&query=nginx.virt)), you have the wrong group configured for access. The fix is simple: Set security.acme.certs.<name>.group to `nginx`, or switch to using the builtins. There's a good example too on [the manual](https://nixos.org/manual/nixos/stable/index.html#module-security-acme-nginx).

m1cr0man commented:
Oh, also, you may need to manually set /var/lib/acme/.challenges to "nginx" group afterwards (since now it already exists, it won't get created with the right permissions). Alternatively delete the folder before you fix the config and the module will recreate it with the right permissions :)

rubenmoor commented:
> Ah of course, I'm just after reading your config thoroughly. Because you've manually configured your certs through nginx instead of using the builtins (like [services.nginx.virtualHosts..enableACME](https://search.nixos.org/options?channel=unstable&show=services.nginx.virtualHosts.%3Cname%3E.enableACME&from=0&size=50&sort=relevance&query=nginx.virt)), you have the wrong group configured for access. The fix is simple: Set security.acme.certs..group to `nginx`, or switch to using the builtins. There's a good example too on [the manual](https://nixos.org/manual/nixos/stable/index.html#module-security-acme-nginx).

Following your suggestion, now the folders all have owner/group set to `acme`/`nginx`. I get a time-out error now instead of 403.

Also, the configuration option `security.acme.certs.<name>.group` is listed in the manual. There is no hint, though, that the default isn't sufficient. The example that explicitly uses `nginx` does not set the option, either.

Also, permissions are set to 754 throughout for all the folders mentioned above.

rubenmoor commented:
Update: the timeout error is based on a different problem in my configuration. During the deployments I have changed the IP at least once, having my domain pointing nowhere. There might have been yet another issue in my security group where I was missing the `.name` in

```
        main.deployment.ec2.securityGroups = [ resources.ec2SecurityGroups.serendipityGroup.name ];
```

By the looks of it, I am stuck with one last issue and about to sort that out myself: I use nginx for ssl certs and my custom server called "backend" to run my app. This custom server is run as a service, thus run by the root user. However, `backend` does not find the ssl key file: "no keys found", even though I point it to the correct directory and file: `/var/lib/acme/staging.rubenmoor.net/key.pem`.

rubenmoor commented:
Finally got my setup to work. Thanks for your advice, @m1cr0man .

Indeed, I needed to set `ecurity.acme.certs.<name>.group` to `nginx`, given my setup.

Ultimately, I had to figure out how to configure `warp-tls` using the chain settings (`tlsSettingsChain` in Haskell)

m1cr0man commented:
> Also, the configuration option `security.acme.certs.<name>.group` is listed in the manual. There is no hint, though, that the default isn't sufficient. The example that explicitly uses `nginx` does not set the option, either.

Glad you got sorted! This error in the docs is fixed in #100356 which hopefully I'll get merged soon.

primeos commented:
FYI: Backport of this PR for NixOS 20.09: #112145 (needs reviewers / not yet merged).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/trying-to-get-a-lets-encrypt-cert-to-renew-connection-refused/11009/2

hmenke commented on nixos/modules/security/acme.nix:
```diff
@@ -248,13 +272,18 @@ let
       # Working directory will be /tmp
       script = ''
-        set -euo pipefail
+        set -euxo pipefail
```
Oh man, is this annoying.

m1cr0man replied:
How so? I added it because debugging people's issues without the shell commands in the logs is quite difficult. Also so much functionality of cert renewal happens in bash. Personally I've found it quite useful.

hmenke replied:
But shouldn't debugging output only be enable when you're, well, debugging?

symphorien replied:
Well with acme, you know you want to debug only after the fact. besides, the amount of logs this generates is not that big...
